torch/_inductor/codegen/cuda/cutlass_utils.py

```diff
@@ -0,0 +1,170 @@
+
+    def __post_init__(self):
+        if self.architectures is None or self.cuda_version is None:
+            raise RuntimeError(
+                f"{self.architectures=} or {self.cuda_version=} is None!"
+            )
+        self._normalize_cuda_arch()
+
+    def _normalize_cuda_arch(self) -> None:
+        assert self.architectures is not None
+        if int(self.architectures) >= 90:
+            self.architectures = "90"
+        elif int(self.architectures) >= 80:
+            self.architectures = "80"
+        elif int(self.architectures) >= 75:
+            self.architectures = "75"
+        elif int(self.architectures) >= 70:
+            self.architectures = "70"
+        else:
+            raise NotImplementedError(f"Unsupported cuda arch: {self.architectures}")
+
+
+@functools.lru_cache(maxsize=1)
+def gen_ops() -> List[Any]:
+    """
+    Generates all supported CUTLASS operations.
+    """
+
+    arch = get_cuda_arch()
+    version = get_cuda_version()
+    if arch is None or version is None:
+        log.error(
+            "Cannot detect cuda arch %s or cuda version %s. "
+            "Will discard all cutlass ops. "
+            "Please consider setting _inductor.cuda.arch and _inductor.cuda.version configs.",
+            arch,
+            version,
+        )
+        return list()
+    args = CUTLASSArgs(architectures=arch, cuda_version=version)
+    manifest = cutlass_manifest.Manifest(args)
+
+    if arch == "90":
+        cutlass_generator.GenerateSM90(manifest, args.cuda_version)
+        cutlass_generator.GenerateSM80(manifest, args.cuda_version)
+    else:
+        try:
+            func = getattr(cutlass_generator, "GenerateSM" + arch)
+            func(manifest, args.cuda_version)
+        except AttributeError as e:
+            raise NotImplementedError(
+                "Arch " + arch + " is not supported by current cutlass lib."
+            ) from e
+
+    return manifest.operations
+
+
+def dtype_match(
+    torch_dtype: torch.dtype, cutlass_dtype: "cutlass_library.DataType"
+) -> bool:
+    if torch_dtype == torch.float:
+        return (
+            cutlass_dtype == cutlass_library.DataType.f32
+            or cutlass_dtype == cutlass_library.DataType.tf32
+        )
+    elif torch_dtype == torch.half:
+        return cutlass_dtype == cutlass_library.DataType.f16
+    elif torch_dtype == torch.bfloat16:
+        return cutlass_dtype == cutlass_library.DataType.bf16
+    else:
+        return False
+
+
+def get_accumulator_dtype(input_torch_dtypes: List[torch.dtype]) -> torch.dtype:
+    """
+    Given a list of input torch dtypes, returns the inferred accumulator torch dtype.
+    """
+
+    if len(input_torch_dtypes) == 0:
+        return None
+    torch_dtype = input_torch_dtypes[0]
+    for dtype in input_torch_dtypes[1:]:
+        if torch_dtype != dtype:
+            raise RuntimeError(f"Unmatched input dtypes: {torch_dtype=}, {dtype=}")
+    if torch_dtype in {torch.float16, torch.half}:
```

Comment:
arent float16 and half the same type? `== torch.float16` might be the same thing?